api/src/main/java/io/jsonwebtoken/SignatureAlgorithm.java

```diff
@@ -576,8 +575,16 @@ public static SignatureAlgorithm forSigningKey(Key key) throws InvalidKeyExcepti
         if (key instanceof SecretKey) {
 
             SecretKey secretKey = (SecretKey)key;
+            String secretKeyAlg = secretKey.getAlgorithm();
             int bitLength = io.jsonwebtoken.lang.Arrays.length(secretKey.getEncoded()) * Byte.SIZE;
 
+            // first, check the key alg name
+            SignatureAlgorithm algFromJcaName = PREFERRED_HMAC_ALGS_LOOKUP.get(secretKeyAlg.toUpperCase(Locale.ENGLISH));
+            if (algFromJcaName != null && bitLength >= algFromJcaName.minKeyLength) {
```

Comment:
I'm not quite sure this is correct.

Only if `bitLength` is `0` *and* you found a `SignatureAlgorithm` from the JCA ID lookup, do you want to short circuit and return the discovered `SignatureAlgorithm` instance.

Otherwise you want to fall back to the `PREFERRED_HMAC_ALGS` priority-based lookup to retain existing behavior.

If we don't want to do it this way, we will probably have to change the JavaDoc to reflect different semantics, and we'd need to think through the repercussions of different semantics.

That is, if someone calls it today and it returns alg A, and they call it after a new release and it returns alg B, that probably wouldn't be a good thing.